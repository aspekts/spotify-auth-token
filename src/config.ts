import { UserConfig } from './index';
import { AppConfig } from './types';
import { goodBye, hasOwnProperty } from './utils';

const defaultConfig: AppConfig = {
  clientId: '',
  clientSecret: '',
  port: 3000,
  outDir: process.cwd(),
  outFileName: 'spotify-token',
  scopes: 'user-read-email',
};

export const createConfig = (args: UserConfig | string[]): AppConfig => {
  const config: AppConfig = { ...defaultConfig };

  // CLI use
  if (Array.isArray(args)) {
    args.forEach((val, idx, orig) => {
      if (val.startsWith('--')) {
        const key = val.slice(2);
        if (hasOwnProperty(config, key)) {
          config[key] = orig[idx + 1];
        }
      }
    });

    // Programmatic use
  } else {
    Object.entries(args).forEach(([key, val]) => {
      if (hasOwnProperty(config, key)) {
        config[key] = val;
      }
    });
  }
  if (!config.clientId || !config.clientSecret) {
    goodBye('Error: Missing client id or client secret');
  }

  if (typeof config.port === 'string') {
    // Convert to number
    config.port = +config.port;
  }

  return config;
};
