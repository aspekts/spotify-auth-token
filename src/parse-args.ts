import path from 'path';
import { SpotifyCreds } from './types';
import { read } from './utils';

export const parseArgs = () => {
  const credentialsFile = process.argv[2];
  if (!credentialsFile) {
    console.error('Error: No credentials file specified');
    process.exit(1);
  }
  const credentialsPath = path.join(process.cwd(), credentialsFile);
  const credentials = read<SpotifyCreds>(credentialsPath);
  if (!credentials.client_id || !credentials.client_secret) {
    console.error('Error: Invalid credentials file');
    process.exit(1);
  }
  return { credentials };
};
