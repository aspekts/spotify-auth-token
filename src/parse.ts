import path from 'path';
import { CLIInputConfig } from './types';
import { read } from './utils';

export const parseArgs = (): CLIInputConfig => {
  const optsFile = process.argv[2];

  if (!optsFile) {
    console.error('Error: No credentials file specified');
    process.exit(1);
  }
  const optsFilePath = path.join(process.cwd(), optsFile);
  const config = read<CLIInputConfig>(optsFilePath);
  if (!config.client_id || !config.client_secret) {
    console.error('Error: Invalid config file');
    process.exit(1);
  }
  return config;
};
