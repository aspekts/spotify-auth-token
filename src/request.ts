import http from 'http';
import https, { RequestOptions } from 'https';

export const getLocalhostUrl = async (port: number): Promise<string> => {
  return new Promise((resolve, reject) => {
    const server = http
      .createServer((req, res) => {
        res.end('You can now close this window');
        res.once('finish', () => {
          server.close();
          if (req.url) {
            resolve(req.url.slice(1));
          }
          reject("Couldn't get code or state");
        });
      })
      .listen(port);
  });
};

export async function request<T>(
  opts: RequestOptions,
  body: string
): Promise<T> {
  return new Promise((resolve, reject) => {
    const data: Uint8Array[] = [];
    const req = https.request(opts, (res) => {
      res.on('data', (chunk) => data.push(chunk));
      res.on('end', () => {
        const resString = Buffer.concat(data).toString();
        resolve(JSON.parse(resString));
      });
    });

    req.on('error', (e) => {
      reject(e);
    });

    req.write(body);
    req.end();
  });
}
