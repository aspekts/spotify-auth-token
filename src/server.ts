import http from 'http';

export const createServerOnce = async (
  port: number,
  callback: () => unknown
): Promise<string> => {
  return new Promise((resolve, reject) => {
    const server = http
      .createServer((req, res) => {
        res.end('You can now close this window');
        res.once('finish', () => {
          server.close();
          if (req.url) {
            resolve(req.url.slice(1));
          }
          reject("Couldn't get code or state");
        });
      })
      .listen(port, undefined, callback);
  });
};
