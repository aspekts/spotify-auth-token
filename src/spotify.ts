import { Config, SpotifyAuthResponse, SpotifyTokenResponse } from './types';
import express from 'express';
import { nanoid } from 'nanoid';
import open from 'open';
import fetch from 'node-fetch';

const app = express();
app.use(express.json());
const state = nanoid(12);

export const getSpotifyToken = async (parms: Config) => {
  const { client_id, client_secret, port, scopes } = parms;

  const server = app.listen(port);
  const spotifyUrl =
    'https://accounts.spotify.com/authorize?' +
    new URLSearchParams({
      response_type: 'code',
      show_dialog: 'true',
      state: encodeURIComponent(state),
      client_id: encodeURIComponent(client_id),
      redirect_uri: `http://localhost:${port}`,
      scope: encodeURIComponent(scopes),
    }).toString();

  await open(spotifyUrl);

  console.info('Please login with Spotify in the browser...');

  const authResonse = await new Promise<SpotifyAuthResponse>(
    (resolve, reject) => {
      app.get('/', (req, res) => {
        res.end('You can now close this window');
        if (req.query.code && req.query.state === state) {
          resolve(req.query as SpotifyAuthResponse);
        } else {
          reject('Something went wrong');
        }
        server.close();
      });
    }
  );

  console.info('Login successfull!');

  const tokenParams = new URLSearchParams({
    grant_type: 'authorization_code',
    code: authResonse.code,
    redirect_uri: 'http://localhost:3000',
  });

  const token: SpotifyTokenResponse = await fetch(
    'https://accounts.spotify.com/api/token',
    {
      body: tokenParams.toString(),
      method: 'POST',
      headers: {
        'Content-type': 'application/x-www-form-urlencoded',
        Authorization:
          'Basic ' +
          Buffer.from(client_id + ':' + client_secret).toString('base64'),
      },
    }
  ).then(res => res.json());

  token.date_obtained = new Date().toLocaleString();
  return token;
};
