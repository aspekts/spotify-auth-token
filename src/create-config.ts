import { defaultConfig } from './config';
import { UserConfig } from './index';
import { AppConfig } from './types';
import { goodBye } from './utils';

function isNumeric(value: string) {
  return /^\d+$/.test(value);
}

export const createConfig = (
  configOrArgs: UserConfig | string[]
): AppConfig => {
  let config: Record<string, string | number> = {};
  // CLI use
  if (Array.isArray(configOrArgs)) {
    config = configOrArgs.reduce((acc, curr, idx, orig) => {
      if (curr.startsWith('--')) {
        const key = curr.slice(2);
        acc[key] = orig[idx + 1];
      }
      return acc;
    }, config);
    // Programmatic use
  } else {
    config = configOrArgs;
  }
  if (!config.clientId || !config.clientSecret) {
    goodBye('Error: Missing client id or client secret');
  }

  if (typeof config.port === 'string') {
    // Convert to number
    config.port = +config.port;
  }

  return { ...defaultConfig, ...config } as AppConfig;
};
