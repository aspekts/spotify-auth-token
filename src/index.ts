import path from 'path';
import { defaultConfig } from './config';
import { getLocalhostUrl, request } from './request';
import type {
  AuthConfig,
  OptionalConfig,
  RequiredConfig,
  SpotifyTokenResponse,
} from './types';
import { id, read, write } from './utils';

export type { AuthConfig };

export default async function run(opts: AuthConfig): Promise<void> {
  try {
    let config = {} as RequiredConfig & OptionalConfig;

    // Programmatic use
    if (opts) {
      config = Object.assign(defaultConfig, opts);

      // CLI us
    } else {
      const usrConfigFilePath = process.argv[2];

      if (!usrConfigFilePath) {
        console.error('Error: No credentials file specified');
        process.exit(1);
      }

      const userConfig: AuthConfig = read(
        path.join(process.cwd(), usrConfigFilePath)
      );

      if (!userConfig.client_id || !userConfig.client_secret) {
        console.error('Error: Invalid config file');
        process.exit(1);
      }
      config = Object.assign(defaultConfig, userConfig);
    }

    const redirectUri = `http://localhost:${config.port}`;
    const state = id();

    const spotifyUrl =
      'https://accounts.spotify.com/authorize?' +
      new URLSearchParams({
        response_type: 'code',
        show_dialog: 'true',
        state: encodeURIComponent(state),
        client_id: encodeURIComponent(config.client_id),
        redirect_uri: redirectUri,
        scope: encodeURIComponent(config.scopes),
      }).toString();

    console.info('Please click the link to login to Spotify in the browser\n');
    console.info(spotifyUrl + '\n');

    const authUrl = await getLocalhostUrl(config.port);
    const params = new URLSearchParams(authUrl);
    const receivedCode = params.get('code');
    const receivedState = params.get('state');

    if (receivedState !== state) {
      throw new Error('Received and original state do not match');
    }

    if (!receivedCode) {
      console.error(receivedCode);
      throw new Error('No code received');
    }

    console.info('Login successfull!');

    const tokenRequestBody = new URLSearchParams({
      grant_type: 'authorization_code',
      code: receivedCode,
      redirect_uri: redirectUri,
    });

    const token: SpotifyTokenResponse = await request(
      {
        hostname: 'accounts.spotify.com',
        path: '/api/token',
        method: 'POST',
        headers: {
          'Content-type': 'application/x-www-form-urlencoded',
          Authorization:
            'Basic ' +
            Buffer.from(config.client_id + ':' + config.client_secret).toString(
              'base64'
            ),
        },
      },
      tokenRequestBody.toString()
    );

    token.date_obtained = new Date().toLocaleString();
    const outDir = path.join(process.cwd(), path.normalize(config.outDir));
    write(outDir, config.outFileName, token);

    console.info('Saved Spotify access token');
    process.exit(0);
  } catch (e) {
    console.error('Something went wrong', e);
    process.exit(1);
  }
}
