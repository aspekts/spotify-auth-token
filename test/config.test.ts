import { createConfig } from '../src/config';

jest.spyOn(process, 'exit').mockImplementation((number) => {
  throw new Error('process.exit: ' + number);
});

describe('createConfig, programmatic use', () => {
  [
    { clientId: 'cid', clientSecret: 'cs' },
    {
      clientId: 'cid',
      clientSecret: 'cs',
      port: 1000,
      scopes: 'scopes',
      outDir: 'secrets',
      outFileName: 'mytoken',
    },
  ].forEach((config, idx) => {
    it(`works #${idx}`, () => {
      expect(createConfig(config)).toMatchSnapshot();
    });
  });

  [
    { clientSecret: 'cs' },
    {
      clientId: 'cid',
    },
  ].forEach((config, idx) => {
    it(`fails #${idx}`, () => {
      expect(() => {
        // @ts-expect-error
        createConfig(config);
      }).toThrow();
    });
  });
});

describe('createConfig, cli use', () => {
  [
    ['--clientId', '111', '--clientSecret', '111'],
    ['--clientId', '333', '--clientSecret', '333', '--port', '4000'],
    ['--clientId', '777', '--clientSecret', '888', '--random', 'nothing'],
    [
      '--clientId',
      '222',
      '--clientSecret',
      '222',
      '--port',
      '1000',
      '--scopes',
      'scopes',
      '--outDir',
      'secrets/spotify',
      '--outFileName',
      'mytoken',
    ],
  ].forEach((config, idx) => {
    it(`works #${idx}`, () => {
      expect(createConfig(config)).toMatchSnapshot();
    });
  });
  [
    ['--clientId', 'cid'],
    ['--clientSecret', 'cs'],
  ].forEach((config, idx) => {
    it(`fails #${idx}`, () => {
      expect(() => {
        createConfig(config);
      }).toThrow();
    });
  });
});
