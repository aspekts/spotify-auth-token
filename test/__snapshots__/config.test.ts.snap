// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createConfig, cli use works #0 1`] = `
Object {
  "clientId": "111",
  "clientSecret": "111",
  "outDir": "",
  "outFileName": "spotify-token",
  "port": 3000,
  "scopes": "user-read-email",
}
`;

exports[`createConfig, cli use works #1 1`] = `
Object {
  "clientId": "333",
  "clientSecret": "333",
  "outDir": "",
  "outFileName": "spotify-token",
  "port": 4000,
  "scopes": "user-read-email",
}
`;

exports[`createConfig, cli use works #2 1`] = `
Object {
  "clientId": "222",
  "clientSecret": "222",
  "outDir": "/secrets",
  "outFileName": "mytoken",
  "port": 1000,
  "scopes": "scopes",
}
`;

exports[`createConfig, programmatic use works #0 1`] = `
Object {
  "clientId": "cid",
  "clientSecret": "cs",
  "outDir": "",
  "outFileName": "spotify-token",
  "port": 3000,
  "scopes": "user-read-email",
}
`;

exports[`createConfig, programmatic use works #1 1`] = `
Object {
  "clientId": "cid",
  "clientSecret": "cs",
  "outDir": "/secrets",
  "outFileName": "mytoken",
  "port": 1000,
  "scopes": "scopes",
}
`;
